test/cpp/microbenchmarks/bm_filters.cc

```diff
@@ -0,0 +1,200 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <benchmark/benchmark.h>
+#include <string.h>
+#include <sstream>
+
+#include <grpc/grpc.h>
+#include <grpc/support/alloc.h>
+#include <grpc/support/string_util.h>
+#include <grpcpp/channel.h>
+#include <grpcpp/support/channel_arguments.h>
+
+#include "src/core/ext/filters/client_channel/client_channel.h"
+#include "src/core/ext/filters/deadline/deadline_filter.h"
+#include "src/core/ext/filters/http/client/http_client_filter.h"
+#include "src/core/ext/filters/http/message_compress/message_compress_filter.h"
+#include "src/core/ext/filters/http/server/http_server_filter.h"
+#include "src/core/ext/filters/load_reporting/server_load_reporting_filter.h"
+#include "src/core/ext/filters/message_size/message_size_filter.h"
+#include "src/core/lib/gprpp/manual_constructor.h"
+
+#include "src/cpp/client/create_channel_internal.h"
+#include "src/proto/grpc/testing/echo.grpc.pb.h"
+#include "test/cpp/microbenchmarks/filter_helpers.h"
+#include "test/cpp/util/test_config.h"
+
+auto& force_library_initialization = Library::get();
+
+// Test a filter's call stack init in isolation. FilterBM, in conjunction with
+// FilterFixture, specifies the filter under test (use the FilterBM<> template
+// to specify this).
+// Note that there is some other work being done within the benchmarking loop,
+// so the result of this microbenchmark is a little bloated.
+template <class FilterBM>
+static void BM_CallStackInit(benchmark::State& state) {
+  // Setup for benchmark
+  FilterBM bm_setup;
+  struct DataForFilterBM data;
+  bm_setup.Setup(&data);
```

Comment:
Could the setup/destroy be handled in the FilterBM constructor/destructor?

Reply:
Why would DataForFilterBM have to live in FilterBM?

Reply:
Because if we're handling setup/destroy in the FilterBM constructor/destructor, then it needs access to the DataForFilterBM object at construction / destruction time

Reply:
Not necessarily:

The ctor could look like:
```
template <class FilterFixture>
class FilterBM {
 public:
  FilterBM(struct DataForFilterBM* data)
      : fixture(), track_counters(), label(), args(), exec_ctx(), factory() {
    data->channel_args = CreateFakeChannelArgs();
    // etc etc etc
  }
};
```
Then, in the benchmark:
```
struct DataForFilterBM data;
FilterBM bm_setup(&data);
```

Reply:
So this way ties in the logic of initializing the struct DataForFilterBM with the constructor of FilterBM. Would this be in convention with how our team treats C++ constructors / destructors? I'm fine with making the change if it doesn't break team convention - it would be kind of nice to have setup and cleanup in constructor and destructor.

Reply:
Actually, now that we are talking style, any reason for the separate struct DataForFilterBM? Why couldn't all of those data be members in FilterBM? It sees like FilterBM logically owns them since it is initializing and destroying them

Reply:
Never mind, Mark linked me to this: https://google.github.io/styleguide/cppguide.html#Access_Control. So I guess public member variable is not an option. What would you suggest?

Reply:
Mark and I talked offline and we think moving the members of DataForFilterBM inside of FilterBM and adding non-const accessors is the best option. Non-const accessors are usually not preferred in non-test code, but they're okay in test code.